File: leetcode3/최원준/3507. Minimum Pair Removal to Sort Array I.py
#

'''
1. 아이디어 :
조건이 n <= 50 이라 n**2으로 풀 수 있다.
정렬일때까지(n):
  최소_sum 찾기(n)

2. 시간복잡도 :
    O(n**2)

3. 자료구조/알고리즘 :
-

'''
class Solution:
    def minimumPairRemoval(self, nums: List[int]) -> int:
        
        def check_sorted(nums: List[int]) -> bool:
            for i in range(len(nums) - 1):
                if nums[i] > nums[i + 1]:
                    return False
            return True
        
        ans = 0

        while not check_sorted(nums):
            min_sum = float("inf")
            min_idx = 0

            for i in range(len(nums) - 1):
                pair_sum = nums[i] + nums[i + 1]

                if pair_sum < min_sum:
                    min_sum = pair_sum
                    min_idx = i

            nums[min_idx] += nums[min_idx + 1]
            nums.pop(min_idx + 1)

            ans += 1

        return ans

File: leetcode3/이진희/3507_Minimum Pair Removal to Sort Array I.java
/*

1. 아이디어 : 
   - 리스트가 오름차순으로 정렬될 때까지 반복
   - 인접한 두 원소의 합이 가장 최소가 되는 위치 찾기
   - 해당 위치의 두 원소를 합치고 카운트 증가

2. 시간복잡도 : O(N^2)

3. 자료구조/알고리즘 : 시뮬레이션

 */

class Solution {
    public int minimumPairRemoval(int[] nums) {
        List<Integer> list = new ArrayList<>();

        for (int num : nums) {
            list.add(num);
        }

        int cnt = 0;

        while (true) {
            boolean sorted = true;

            for (int i = 0; i < list.size() - 1; i++) {
                if (list.get(i) > list.get(i + 1)) {
                    sorted = false;
                    break;
                }
            }

            if (sorted) {
                break;
            }

            int minIndex = 0;
            int minSum = list.get(0) + list.get(1);

            for (int i = 1; i < list.size() - 1; i++) {
                int sum = list.get(i) + list.get(i + 1);

                if (sum < minSum) {
                    minSum = sum;
                    minIndex = i;
                }
            }

            list.set(minIndex, minSum);
            list.remove(minIndex + 1);

            cnt++;
        }

        return cnt;
    }
}